helix-term/src/commands.rs

```diff
@@ -3647,6 +3648,36 @@ fn yank(cx: &mut Context) {
     exit_select_mode(cx);
 }
 
+fn yank_joined_impl(editor: &mut Editor, separator: &str, register: char) -> anyhow::Result<()> {
+    let (view, doc) = current!(editor);
+    let text = doc.text().slice(..);
+
+    let mut slices = doc.selection(view.id).slices(text);
+    let init = slices.next().unwrap().to_string();
+    let (num_selections, joined) = slices.fold((1, init), |(n, mut acc), slice| {
+        acc.push_str(separator);
+        acc.push_str(slice.as_str().unwrap());
+        (n + 1, acc)
+    });
```

Comment:
A few things here:

* we can get the number of selections up-front with `Selection::len`
* we probably want to iterate over `Selection::fragments` since the Cow will be nicer to work with than the `Option<&str>` from slices
* there's a pattern in a few places we have for joining a string with a separator using `fold` and an `is_empty` check

```rust
let selection = doc.selection(view.id);
let selections = selection.len();

let joined = selection
    .fragments(text)
    .fold(String::new(), |mut acc, fragment| {
        if !acc.is_empty() {
            acc.push_str(separator);
        }
        acc.push_str(&fragment);
        acc
    });
```